### ISSUE TITLE
Subscribe Patterns does not work

### ISSUE THREAD

**Issue by Sineos** [ioBroker/ioBroker.mqtt#24]
**Description**

Adapter version: 2.0.3

Using the adapter as Client / Subscriber. The `Subscribe Patterns` is either set to empty or to a not existing topic. 
MQTT Server is Mosquitto on a remote host.

**Actual behavior:**

The adapter subscribes to every topic published by Mosquitto.

**Expected behavior**

Do not subscribe to any topic.

**Comment by zapccu** [comment]
Problem is: Subscription is made before client is connected to broker. After connection is established, subscription is reset to all topics.

Solution: Subscribe topics in Handler client.on('Connect')

**Comment by Apollon77** [comment]
please check 2.0.4 from GitHub and give feedback

**Comment by Sineos** [comment]
Many thanks for working on it. Unfortunately 2.0.4 does not solve it for me, I entered `none` as pattern and the adapter still subscribes to all topics published. Same is true if leaving the pattern empty.

**Comment by Apollon77** [comment]
Ok, sorry I missed to tell what I changed ... it was more pointed to @zapccu ... I moved the point where subscritions take place.

As soon as you enter something in patterns it should be used and only subscribed to this.

What is shown in logs? Please set to debug anf send the log

**Comment by Sineos** [comment]
![grafik](https://user-images.githubusercontent.com/4459312/49334036-4a637b80-f5cb-11e8-8e9b-28c317a7a60b.png)

Worflow:
- Stop mqtt.0 instance
- Set to "silly"
- Delete entire mqtt.0 Object group
- Start mqtt.0

Log:
```
mqtt.0	2018-12-02 00:40:55.759	silly	redis publish io.mqtt.0.value_only.power.sml.Bezug {"val":1039.8613518198,"ack":true,"ts":1543707655759,"q":0,"from":"system.adapter.mqtt.0","lc":1543707655759}
mqtt.0	2018-12-02 00:40:55.759	debug	Create object for topic: value_only/power/sml/Bezug[ID: mqtt.0.value_only.power.sml.Bezug]
mqtt.0	2018-12-02 00:40:55.752	silly	redis publish io.mqtt.0.power.sml.Bezug {"val":"raw_Bezug value=1039.8613518198 1543707655696000000","ack":true,"ts":1543707655752,"q":0,"from":"system.adapter.mqtt.0","lc":1543707655752}
mqtt.0	2018-12-02 00:40:55.752	debug	Create object for topic: power/sml/Bezug[ID: mqtt.0.power.sml.Bezug]
mqtt.0	2018-12-02 00:40:55.740	silly	redis publish io.mqtt.0.value_only.power.sml.Leistung {"val":1039.8,"ack":true,"ts":1543707655738,"q":0,"from":"system.adapter.mqtt.0","lc":1543707655738}
mqtt.0	2018-12-02 00:40:55.740	silly	redis publish io.mqtt.0.power.sml.Leistung {"val":"raw_Leistung value=1039.8 1543707655696000000","ack":true,"ts":1543707655737,"q":0,"from":"system.adapter.mqtt.0","lc":1543707655737}
mqtt.0	2018-12-02 00:40:55.740	debug	Create object for topic: value_only/power/sml/Leistung[ID: mqtt.0.value_only.power.sml.Leistung]
mqtt.0	2018-12-02 00:40:55.740	debug	Create object for topic: power/sml/Leistung[ID: mqtt.0.power.sml.Leistung]
mqtt.0	2018-12-01 23:40:54.583	silly	redis publish io.mqtt.0.info.connection {"val":true,"ack":true,"ts":1543707654582,"q":0,"from":"system.adapter.mqtt.0","lc":1543707654579}
mqtt.0	2018-12-01 23:40:54.579	silly	redis publish io.mqtt.0.info.connection {"val":true,"ack":true,"ts":1543707654579,"q":0,"from":"system.adapter.mqtt.0","lc":1543707654579}
mqtt.0	2018-12-01 23:40:54.577	info	Subscribe on: "none/#"
mqtt.0	2018-12-01 23:40:54.577	info	Connected to 192.168.178.8
mqtt.0	2018-12-01 23:40:54.526	info	Try to connect to mqtt://192.168.178.8:1883?clientId=iobroker.mqtt.0
```

`power.sml.Leistung` and all other topics mentioned in the log snippet are not native to the ioBroker instance and are only published by a Mosquitto instance running on a different IP.

Basically it is exactly what my first post or @zapccu describes: Although subscription is set to the not existing topic `none/#`, the adapter grabs everything that is published by the broker.

**Comment by Apollon77** [comment]
But as you can see a subscribe is only set for none/# ... I thought the MQTT client only work with what it subscribed to. In this case it is not subscribed ... Strange

**Comment by Apollon77** [comment]
Can you please enter soemthing ith our the /# at the end please?

**Comment by Sineos** [comment]
In my tests it made no difference if I set the pattern to `none`, `none/#` or leaving it empty.

```
mqtt.0 | 2018-12-02 10:54:28.269 | silly | redis publish io.mqtt.0.power.sml.Leistung {"val":"raw_Leistung  value=2998.1  1543748068243000000","ack":true,"ts":1543748068269,"q":0,"from":"system.adapter.mqtt.0","lc":1543748068269}
mqtt.0 | 2018-12-02 10:54:28.267 | debug | Create object for topic: power/sml/Leistung[ID: mqtt.0.power.sml.Leistung]
mqtt.0 | 2018-12-02 10:54:27.491 | silly | redis publish io.mqtt.0.power.pv.consumption {"val":"raw_PVconsumption  value=3228  1543748055000000000","ack":true,"ts":1543748067491,"q":0,"from":"system.adapter.mqtt.0","lc":1543748067491}
mqtt.0 | 2018-12-02 10:54:27.489 | debug | Create object for topic: power/pv/consumption[ID: mqtt.0.power.pv.consumption]
mqtt.0 | 2018-12-02 10:54:27.029 | silly | redis publish io.mqtt.0.info.connection  {"val":true,"ack":true,"ts":1543748067029,"q":0,"from":"system.adapter.mqtt.0","lc":1543748067029}
mqtt.0 | 2018-12-02 10:54:27.027 | info | Subscribe on: "none"
mqtt.0 | 2018-12-02 10:54:27.027 | info | Connected to 192.168.178.8
mqtt.0 | 2018-12-02 10:54:27.023 | silly | redis publish io.mqtt.0.info.connection  {"val":false,"ack":true,"ts":1543748067023,"q":0,"from":"system.adapter.mqtt.0","lc":1543748067023}
mqtt.0 | 2018-12-02 10:54:26.977 | info | Try to connect to mqtt://192.168.178.8:1883?clientId=iobroker.mqtt.0
mqtt.0 | 2018-12-02 10:54:26.940 | silly | redis mget 1 1
mqtt.0 | 2018-12-02 10:54:26.938 | silly | redis psubscribe io.mihome-airpurifier.0.info.humidity
mqtt.0 | 2018-12-02 10:54:26.918 | silly | redis mget 1 1
mqtt.0 | 2018-12-02 10:54:26.916 | silly | redis publish io.system.adapter.mqtt.0.outputCount  {"val":8,"ack":true,"ts":1543748066916,"q":0,"from":"system.adapter.mqtt.0","lc":1543748066916}
mqtt.0 | 2018-12-02 10:54:26.916 | silly | redis publish io.system.adapter.mqtt.0.inputCount  {"val":0,"ack":true,"ts":1543748066916,"q":0,"from":"system.adapter.mqtt.0","lc":1543707894504}
mqtt.0 | 2018-12-02 10:54:26.916 | silly | redis publish io.system.adapter.mqtt.0.uptime  {"val":0,"ack":true,"ts":1543748066916,"q":0,"from":"system.adapter.mqtt.0","lc":1543748066916}
mqtt.0 | 2018-12-02 10:54:26.916 | silly | redis publish io.system.adapter.mqtt.0.memHeapUsed  {"val":14.62,"ack":true,"ts":1543748066916,"q":0,"from":"system.adapter.mqtt.0","lc":1543748066916}
mqtt.0 | 2018-12-02 10:54:26.915 | silly | redis publish io.system.adapter.mqtt.0.memHeapTotal  {"val":29.83,"ack":true,"ts":1543748066915,"q":0,"from":"system.adapter.mqtt.0","lc":1543748066915}
mqtt.0 | 2018-12-02 10:54:26.915 | silly | redis publish io.system.adapter.mqtt.0.memRss  {"val":46.26,"ack":true,"ts":1543748066915,"q":0,"from":"system.adapter.mqtt.0","lc":1543748066915}
mqtt.0 | 2018-12-02 10:54:26.915 | silly | redis publish io.system.adapter.mqtt.0.connected  {"val":true,"ack":true,"ts":1543748066915,"q":0,"from":"system.adapter.mqtt.0","lc":1543748066915}
mqtt.0 | 2018-12-02 10:54:26.915 | silly | redis publish io.system.adapter.mqtt.0.alive  {"val":true,"ack":true,"ts":1543748066915,"q":0,"from":"system.adapter.mqtt.0","lc":1543748066915}
mqtt.0 | 2018-12-02 10:54:26.914 | silly | redis keys 1 *.logging
mqtt.0 | 2018-12-02 10:54:26.913 | silly | redis psubscribe io.mqtt.0.checkLogging
mqtt.0 | 2018-12-02 10:54:26.911 | info | starting. Version 2.0.4 in /opt/iobroker/node_modules/iobroker.mqtt, node: v8.12.0
```

Maybe just a question of wording, but still something that made me wonder:

`mqtt.0 | 2018-12-02 10:54:26.938 | silly | redis psubscribe io.mihome-airpurifier.0.info.humidity`

In fact this is a local ioBroker state, that I set to being **published** by the mqtt adapter (see screenshot in my post above) --> Log states `redis psubscribe`

`mqtt.0 | 2018-12-02 10:54:27.491 | silly | redis publish io.mqtt.0.power.pv.consumption {"val":"raw_PVconsumption  value=3228`

This is an external value being published by Mosquitto broker --> Log states `redis publish`

**Comment by Apollon77** [comment]
set logging to debug instead silly (as told) and all this internal redis logs disppear.

redis.psubscribe only means. that the adapter wnts change infos for this ionroker object. redis publish means that adapter write info iobroker internally!!

All those "redis..." logs have NOTHING to do with mqtt!!

**Comment by Apollon77** [comment]
... at least not directly. But yes all mqtt-incoming data will be shown as "redis publish" when written into iobroker

**Comment by Sineos** [comment]
> set logging to debug instead silly (as told) and all this internal redis logs disppear.

The reason for "silly" is quite simple: You will not see a single incoming mqtt topic when set to "debug". So to clarify the issue here, that in fact the mqtt adapter is subscribing to topics where it should not, it makes sense to use silly. 
That the redis messages are not directly related to this issue, is clear.

**Comment by zapccu** [comment]
Installed 2.0.4. Works. Thank you for fixing it!

**Comment by Sineos** [comment]
I just cross checked with [MQTT.fx](https://www.jensd.de/wordpress/) and subscribing / unsubscribing works perfectly fine.

Nevertheless I have solved the issue: It appears when the Mosquitto broker has the setting `persistence true`. Then it seems that the adapter is no longer able to correctly subscribe to the topics (whereas MQTT.fx has no issue with this setting)

Many thanks for your support. Appreciated.

**Comment by Apollon77** [comment]
Great! Then unpublished update later and close here

**Comment by Sineos** [comment]
> Nevertheless I have solved the issue: It appears when the Mosquitto broker has the setting persistence true. Then it seems that the adapter is no longer able to correctly subscribe to the topics (whereas MQTT.fx has no issue with this setting)

I have to revert this statement, the problem still persists.

Reason:
- The adapter is setting `clean session` to false as default
- This means the adapter actively requests a persistent session from the broker

Effect:
When the adapter is started with a default `#` subscription and the subscription is change later on, the `#` "survives" the adapter restart as it is persisted in the broker.

#27 is a take on fixing this: It sets the default to clean session `true`, which in my opinion fits most common use cases of the adapter and it exposes the clean session flag as a setting in the adapter's config.

**Comment by Sineos** [comment]
Closed with #27. Thanks
